### ISSUE TITLE
Generated TypeAdapterFactory and null safety

### ISSUE THREAD

**Issue by muddin1** [rharter/auto-value-gson#50]
As of 0.2.5, My custom AutoValueTypeAdapterFactory would call `nullSafe()` on every type adapter I return in the `create()` method. Refer https://github.com/rharter/auto-value-gson/issues/28#issuecomment-211680677.

``` java
Class<? super T> rawType = type.getRawType();
if (rawType.equals(User.class)) {
    return (TypeAdapter<T>) User.typeAdapter(gson).nullSafe();
}
```

Beginning from 0.3.0, the generated AutoValueGsonTypeAdapterFactory does not call `nullSafe()` by default, which is the appropriate way for sure.

So, how'd I go about ensuring null-safety when using the generated TypeAdapterFactory.

**Comment by rharter** [comment]
You could submit a PR if you've got the code.  Otherwise I'll have to add that.

**Comment by muddin1** [comment]
Are you saying that we should attach `nullSafe()` to every AutoValue typeAdapter mentioned in the generated Factory i.e. in the annotation processor?

Also, there's this PR https://github.com/rharter/auto-value-gson/pull/23 that presumably eliminates the need to use `nullSafe()` completely.

**Comment by jachenry** [comment]
Below is a quick solution for anyone else that hits this before #23 is completed.

```java
@GsonTypeAdapterFactory
public abstract class AutoValueAdapterFactory implements TypeAdapterFactory {

  public static TypeAdapterFactory create() {
    final TypeAdapterFactory factory = new AutoValueGson_AutoValueAdapterFactory();

    return new TypeAdapterFactory() {
      @Override
      public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
        TypeAdapter<T> typeAdapter = factory.create(gson, type);
        return typeAdapter != null ? typeAdapter.nullSafe() : null;
      }
    };
  }
}
```

**Comment by muddin1** [comment]
@jachenry Thanks, I've been doing the same.

**Comment by jjimenez0611** [comment]
Hi @jachenry 

Please can you explain to me with more details..

I have my factory like this:

public class **AutoValueGsonTyeAdapterFactory** implements TypeAdapterFactory  {

    public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
        Class<? super T> rawType = type.getRawType();

        if (rawType.equals(MGuest.class)) {
            return (TypeAdapter<T>) MGuest.typeAdapter(gson);
        }

        if(rawType.equals(MTicket.class)){
            return (TypeAdapter<T>) MTicket.typeAdapter(gson);
        }

        return null;
    }
}

**Then I have your abstract class that contains my class**

@GsonTypeAdapterFactory
public abstract class AutoValueAdapterFactory implements TypeAdapterFactory {

    public static TypeAdapterFactory create() {
        final TypeAdapterFactory factory = new **AutoValueGsonTyeAdapterFactory();**

        return new TypeAdapterFactory() {
            @Override
            public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
                TypeAdapter<T> typeAdapter = factory.create(gson, type);
                return typeAdapter != null ? typeAdapter.nullSafe() : null;
            }
        };
    }
}


And finally I need to call this classes in my GsonConverterFactory, but in this point I really don't now how to call the abstract class... 


public static GsonConverterFactory gsonConverterFactory = GsonConverterFactory.create(
            new GsonBuilder()
                    .registerTypeAdapterFactory(**MyTypeAdapterFactory**)
                    .create());


Can you help me in this last step?

**Comment by jachenry** [comment]
@jjimenez0611 You'll want to call the static `create()` method on `AutoValueAdapterFactory`.

```
@GsonTypeAdapterFactory
public abstract class AutoValueAdapterFactory implements TypeAdapterFactory {

public static TypeAdapterFactory create() {
    final TypeAdapterFactory factory = new AutoValueGson_AutoValueAdapterFactory();

    return new TypeAdapterFactory() {
        @Override
        public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
            TypeAdapter<T> typeAdapter = factory.create(gson, type);
            return typeAdapter != null ? typeAdapter.nullSafe() : null;
        }
    };
}
```

```
new GsonBuilder()
    .registerTypeAdapterFactory(AutoValueAdapterFactory.create())
    .create()
```

**Comment by jjimenez0611** [comment]
@jachenry, thanks I really appreciate your help, now I understand.

**Comment by hzsweers** [comment]
Closed by #78
